packages/core/event.js

```diff
@@ -126,33 +127,130 @@ const stringOrFallback = (str, fallback) => typeof str === 'string' && str ? str
-        (frame.functionName || '').indexOf('StackGenerator$$') === -1
-      ).slice(1 + generatedFramesToSkip)
-    } catch (e) {
-      return []
-    }
+    return []
   }
 }
 
-BugsnagEvent.ensureEvent = function (eventOrError, errorFramesToSkip = 0, generatedFramesToSkip = 0) {
-  // notify() can be called with an Event object. In this case no action is required
-  if (eventOrError.__isBugsnagEvent) return eventOrError
+BugsnagEvent.create = function (maybeError, tolerateNonErrors, handledState, component, errorFramesToSkip = 0, logger) {
+  const [error, internalFrames] = normaliseError(maybeError, tolerateNonErrors, component, logger)
+  let event
   try {
-    const stacktrace = BugsnagEvent.getStacktrace(eventOrError, errorFramesToSkip, 1 + generatedFramesToSkip)
-    return new BugsnagEvent(eventOrError.name, eventOrError.message, stacktrace, undefined, eventOrError)
+    const stacktrace = BugsnagEvent.getStacktrace(
+      error,
+      // if an error was created/throw in the normaliseError() function, we need to
+      // tell the getStacktrace() function to skip the number of frames we know will
+      // be from our own functions. This is added to the number of frames deep we
+      // were told about
+      internalFrames > 0 ? 1 + internalFrames + errorFramesToSkip : 0,
+      // if there's no stacktrace, the callstack may be walked to generated one.
+      // this is how many frames should be removed because they come from our library
+      1 + errorFramesToSkip
+    )
+    event = new BugsnagEvent(error.name, error.message, stacktrace, handledState, maybeError)
   } catch (e) {
-    return new BugsnagEvent(eventOrError.name, eventOrError.message, [], undefined, eventOrError)
+    event = new BugsnagEvent(error.name, error.message, [], handledState, maybeError)
+  }
+  if (error.name === 'InvalidError') {
+    event.addMetadata(`${component}`, 'non-error parameter', makeSerialisable(maybeError))
+  }
+  return event
+}
+
+const makeSerialisable = (err) => {
+  if (err === null) {
+    return 'null'
+  } else if (err === undefined) {
+    return 'undefined'
+  } else if (isError(err)) {
+    return {
+      [Object.prototype.toString.call(err)]: {
+        name: err.name,
+        message: err.message,
+        code: err.code,
+        stack: err.stack
+      }
+    }
+  } else {
+    return err
   }
 }
 
+const normaliseError = (maybeError, tolerateNonErrors, component, logger) => {
+  let error
+  let internalFrames = 0
+
+  const createAndLogInputError = (reason, value) => {
+    if (logger) logger.warn(`${component} received a non-error: "${reason}"`)
+    const err = new Error(`${component} received a non-error. See "${component}" tab for more detail.`)
+    err.name = 'InvalidError'
+    return err
+  }
+
+  if (!tolerateNonErrors) {
+    if (isError(maybeError)) {
+      error = maybeError
+    } else {
+      error = createAndLogInputError(typeof maybeError, maybeError)
+      internalFrames += 2
```

Comment:
The +2 is because when know the stacktrace will include a frame from `normaliseError` and `createAndLogInputError` at this point.